File: nightingale/androidApp/app/src/main/java/net/acumenus/nightingale/NightingalePresentationPreferences.kt
package net.acumenus.nightingale

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

internal object NightingalePresentationPreferenceNamespace {
    const val PREFERENCES_FILE = "net.acumenus.nightingale.presentation.v1"
    const val REDUCE_MOTION = "reduce-motion"
    const val HIDE_DECORATIVE_IMAGERY = "hide-decorative-imagery"

    val allKeys = setOf(REDUCE_MOTION, HIDE_DECORATIVE_IMAGERY)
}

internal data class NightingalePresentationPreferenceSnapshot(
    val reduceMotionRequested: Boolean,
    val hideDecorativeImageryRequested: Boolean,
)

internal data class NightingaleSceneAccessibilityPolicy(
    val reduceMotion: Boolean,
    val showDecorativeImagery: Boolean,
    val imageAlpha: Float,
    val scrimAlphas: List<Float>,
    val transitionDurationMillis: Int,
)

internal fun nightingaleSceneAccessibilityPolicy(
    preferences: NightingalePresentationPreferenceSnapshot,
    fontScale: Float,
    highContrast: Boolean,
    systemReduceMotion: Boolean,
): NightingaleSceneAccessibilityPolicy {
    val reduceMotion = systemReduceMotion || preferences.reduceMotionRequested
    val showDecorativeImagery =
        !preferences.hideDecorativeImageryRequested && !highContrast
    val imageAlpha = when {
        !showDecorativeImagery -> 0f
        fontScale >= 1.3f -> 0.08f
        else -> 0.16f
    }
    val scrimAlphas = when {
        highContrast -> listOf(1f, 1f, 1f)
        fontScale >= 1.3f -> listOf(0.92f, 0.96f, 0.99f)
        else -> listOf(0.72f, 0.86f, 0.96f)
    }

    return NightingaleSceneAccessibilityPolicy(
        reduceMotion = reduceMotion,
        showDecorativeImagery = showDecorativeImagery,
        imageAlpha = imageAlpha,
        scrimAlphas = scrimAlphas,
        transitionDurationMillis = if (reduceMotion) 0 else 180,
    )
}

internal class NightingalePresentationPreferences(context: Context) {
    private val preferences = context.getSharedPreferences(
        NightingalePresentationPreferenceNamespace.PREFERENCES_FILE,
        Context.MODE_PRIVATE,
    )

    var snapshot by mutableStateOf(readSnapshot())
        private set

    fun setReduceMotionRequested(requested: Boolean): Boolean {
        val committed = preferences.edit()
            .putBoolean(
                NightingalePresentationPreferenceNamespace.REDUCE_MOTION,
                requested,
            )
            .commit()
        if (committed) {
            snapshot = snapshot.copy(reduceMotionRequested = requested)
        }
        return committed
    }

    fun setHideDecorativeImageryRequested(requested: Boolean): Boolean {
        val committed = preferences.edit()
            .putBoolean(
                NightingalePresentationPreferenceNamespace.HIDE_DECORATIVE_IMAGERY,
                requested,
            )
            .commit()
        if (committed) {
            snapshot = snapshot.copy(hideDecorativeImageryRequested = requested)
        }
        return committed
    }

    private fun readSnapshot() = NightingalePresentationPreferenceSnapshot(
        reduceMotionRequested = preferences.getBoolean(
            NightingalePresentationPreferenceNamespace.REDUCE_MOTION,
            false,
        ),
        hideDecorativeImageryRequested = preferences.getBoolean(
            NightingalePresentationPreferenceNamespace.HIDE_DECORATIVE_IMAGERY,
            false,
        ),
    )
}

File: nightingale/iosApp/Nightingale/NightingalePresentationPreferences.swift
import Combine
import Foundation

enum NightingalePresentationPreferenceNamespace {
    static let reduceMotion = "net.acumenus.nightingale.presentation.v1.reduce-motion"
    static let hideDecorativeImagery =
        "net.acumenus.nightingale.presentation.v1.hide-decorative-imagery"

    static let allKeys = [
        reduceMotion,
        hideDecorativeImagery,
    ]
}

struct NightingalePresentationPreferenceSnapshot: Equatable {
    let reduceMotionRequested: Bool
    let hideDecorativeImageryRequested: Bool
}

struct NightingaleSceneAccessibilityPolicy: Equatable {
    let reduceMotion: Bool
    let showDecorativeImagery: Bool
    let decorativeImageOpacity: Double
    let backgroundScrimAlphas: [Double]
    let cardOpacity: Double
    let transitionDuration: Double

    static func resolve(
        preferences: NightingalePresentationPreferenceSnapshot,
        systemReduceMotion: Bool,
        systemReduceTransparency: Bool,
        increasedContrast: Bool,
        accessibilityTextSize: Bool,
        darkMode: Bool
    ) -> NightingaleSceneAccessibilityPolicy {
        let reduceMotion = systemReduceMotion || preferences.reduceMotionRequested
        let showDecorativeImagery =
            !preferences.hideDecorativeImageryRequested
            && !systemReduceTransparency
            && !increasedContrast

        let imageOpacity: Double
        let backgroundScrimAlphas: [Double]
        if !showDecorativeImagery {
            imageOpacity = 0
            backgroundScrimAlphas = [1, 1, 1]
        } else if accessibilityTextSize {
            imageOpacity = 1
            backgroundScrimAlphas = [0.72, 0.88, 0.97]
        } else if darkMode {
            imageOpacity = 1
            backgroundScrimAlphas = [0.58, 0.78, 0.92]
        } else {
            imageOpacity = 1
            backgroundScrimAlphas = [0.46, 0.70, 0.88]
        }

        return NightingaleSceneAccessibilityPolicy(
            reduceMotion: reduceMotion,
            showDecorativeImagery: showDecorativeImagery,
            decorativeImageOpacity: imageOpacity,
            backgroundScrimAlphas: backgroundScrimAlphas,
            cardOpacity: systemReduceTransparency || increasedContrast ? 1 : 0.96,
            transitionDuration: reduceMotion ? 0 : 0.18
        )
    }
}

@MainActor
final class NightingalePresentationPreferences: ObservableObject {
    @Published private(set) var snapshot: NightingalePresentationPreferenceSnapshot

    private let defaults: UserDefaults

    init(
        defaults: UserDefaults = .standard,
        environment: [String: String] = ProcessInfo.processInfo.environment
    ) {
        self.defaults = defaults

        #if DEBUG
        if environment["NIGHTINGALE_TEST_RESET_PRESENTATION_PREFERENCES"] == "1" {
            for key in NightingalePresentationPreferenceNamespace.allKeys {
                defaults.removeObject(forKey: key)
            }
        }
        #endif

        snapshot = NightingalePresentationPreferenceSnapshot(
            reduceMotionRequested: defaults.bool(
                forKey: NightingalePresentationPreferenceNamespace.reduceMotion
            ),
            hideDecorativeImageryRequested: defaults.bool(
                forKey: NightingalePresentationPreferenceNamespace.hideDecorativeImagery
            )
        )
    }

    func setReduceMotionRequested(_ requested: Bool) {
        defaults.set(
            requested,
            forKey: NightingalePresentationPreferenceNamespace.reduceMotion
        )
        snapshot = NightingalePresentationPreferenceSnapshot(
            reduceMotionRequested: requested,
            hideDecorativeImageryRequested: snapshot.hideDecorativeImageryRequested
        )
    }

    func setHideDecorativeImageryRequested(_ requested: Bool) {
        defaults.set(
            requested,
            forKey: NightingalePresentationPreferenceNamespace.hideDecorativeImagery
        )
        snapshot = NightingalePresentationPreferenceSnapshot(
            reduceMotionRequested: snapshot.reduceMotionRequested,
            hideDecorativeImageryRequested: requested
        )
    }
}
